In [1]:
import torch
from tinypytorch.data import get_local_data
from tinypytorch.model import initialize_parameters, Model

In [2]:
#| hide
from tinypytorch.core import *

# tinypytorch

> A tiny version of pytorch from scratch

This file will become your README and also the index of your documentation.

## Install

```sh
pip install tinypytorch
```

### Train a neural network

#### Data

In [3]:
x_train, y_train, x_valid, y_valid = get_local_data()

In [4]:
x_train.shape, y_train.shape

(torch.Size([50000, 784]), torch.Size([50000]))

In [5]:
y_train[1]

tensor(0)

In [6]:
y_train[1:5]

tensor([0, 4, 1, 9])

#### Initialize hyperparameters

In [7]:
n, m = x_train.shape # num rows and columns

In [8]:
c = y_train.max() + 1

In [9]:
n, m, c

(50000, 784, tensor(10))

In [10]:
nh = 50 # num hidden

In [11]:
w1, b1, w2, b2 = initialize_parameters(m, nh)

In [12]:
w1.shape, b1.shape

(torch.Size([784, 50]), torch.Size([50]))

In [13]:
w2.shape, b2.shape

(torch.Size([50, 1]), torch.Size([1]))

- Training set's shape: (50000, 784)
- Weight's shape: (784, 50)
- Bias's shape: (50)

#### The first layer (Lin): (50000, 784) x (784, 50) + (50)

In [14]:
model = Model(w1, b1, w2, b2)

In [15]:
loss = model(x_train, y_train)

Model.__call__
l=<tinypytorch.model.Lin object at 0x7fcb20bb4610>
x.shape=torch.Size([50000, 50])
Model.__call__
l=<tinypytorch.model.ReLU object at 0x7fcb20bb4310>
x.shape=torch.Size([50000, 50])
Model.__call__
l=<tinypytorch.model.Lin object at 0x7fcb20bb42e0>
x.shape=torch.Size([50000, 1])
MSE.forward
inp.shape=torch.Size([50000, 1])
inp.squeeze().shape=torch.Size([50000])
targ.shape=torch.Size([50000])
temp=tensor([-4.4184,  0.4608, -3.0481,  ..., -7.5595, -3.5340, -7.3656])


In [16]:
loss

tensor(22.0779)

In [17]:
model.backward()